tpl/tags/tags.go

```diff
@@ -0,0 +1,95 @@
+package tags
+
+import (
+	"fmt"
+	"html/template"
+
+	"github.com/gohugoio/hugo/deps"
+)
+
+// New creates a pointer to a new tags.Namespace struct.
+func New(d *deps.Deps) *Namespace {
+	return &Namespace{d}
+}
+
+type Namespace struct {
+	// how to get the source folder path info (e.g. --source YYY).
+	*deps.Deps
+}
+
+// Script generates a script tag and concatenates the srcs into dest.
+//
+func (ns *Namespace) Script(dest string, attr map[string]interface{}, srcs ...interface{}) (template.HTML, error) {
+	sourceFs := ns.Fs.WorkingDir
+	destinationFs := ns.Fs.Destination
+	var html template.HTML
+
+	bytes, err := ReadBytes(sourceFs, srcs...)
+	if err != nil {
+		return html, err
+	}
+
+	asset := NewAsset(dest, bytes)
+	url := asset.Url(true)
+
+	// TODO (nfisher 2018-01-25): Might need to introduce a write lock to prevent partial/interleaved writes?
+	err = Prepare(destinationFs, ns.PublishDir, dest)
+	if err != nil {
+		return html, err
+	}
+
+	err = Write(destinationFs, ns.PublishDir, url, bytes)
```

Comment:
@bep what kind of concurrency do I have to worry about here with Hugo? Does afero provide "file locks"? I could push this on a channel and have a single go routine that writes files out and dedups where required.

Reply:
I think we need to discuss the issue before discussing implementation.

Reply:
Quick background these tags;

 a) generate new files (CSS or JS) via concatenation where multiple SRC are specified so that the number of requests can be reduced where desired.
 b) copy and rename the file contents using DEST as a base and appending the shasum before the file extension for cache-busting purposes.
 c) add a SRI integrity attribute to the resources HTML tag for some measure of increased security where CDN's are used.

The primary motivation for this PR is to minimise hand cranking the SRI and cache-busting tags for every change which are both based on the files shasum/identity. If an integrity attributes shasum doesn't match the shasum of it's contents modern browsers will not load that file. (It's typically employed in combination with CDN's as a security measure.)

I'm currently trying to understanding how Hugo processes each "post" and it's associated templates. I'm assuming it's in parallel. As a result I suspect theres potential data races for who's seen the asset first. That race involves two elements that would require some form of synchronisation/mutual exclusion to ensure correct behaviour;

1. caching shasum calculations.
2. writing to DEST.

The issue I'm concerned with is how to handle one of these functions being placed in a baseof template as an example. If I have a site generating 10s or 100s of posts presumably the baseof template will be evaluated for each post. Cached shasums is easy to handle, stuff it in a lockable map. The bit I'm less certain about is writing the files to disk, currently I'm naively writing the files on every invocation of these functions. I don't see this as optimal because it's inefficient and could lead to data corruption assuming it's possible to have the same file opened by multiple writers, potentially resulting in interleaved writes and corrupt data.